### PR TITLE
Feature: Add support for choosing the default sorting and grouping options

### LINKED ISSUE #7082: Feature: Add option to set default sorting options
### Requirements
- Add combo box drop down to folder settings page similar to the default layout options
   - Settings can still be overridden on individual directories 

### Files Version

Version: 2.0.34.0

### Windows Version

 OS Version: 10.0.19044.1348

### Additional comment

_No response_

### PR DESCRIPTION
**Pull request by manfromarce**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #7082

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

### CODE REVIEW

**Comment by yaira2** [comment]
Locking this PR as the feature request requirements haven't been finalized. Let's continue dicussing in the issue.

**Comment by yaira2** [comment]
Requirements have been sorted out

**Comment by manfromarce** [comment]
Ok, if I understand correctly I need to remove the separate "override sort settings" option and use the existing Override layout option, and move the new settings to the Folders tab. Should I also move "Sort directories alongside files" which was already in Preferences?

**Comment by yaira2** [comment]
> use the existing Override layout option, and move the new settings to the Folders tab.

Correct

> Should I also move "Sort directories alongside files" which was already in Preferences?

Let's leave as is for now

File: src/Files.App/ViewModels/FolderSettingsViewModel.cs [hunk 1]
```diff
@@ -287,47 +287,59 @@ public ColumnsViewModel ColumnsViewModel
 		private static LayoutPreferences GetLayoutPreferencesForPath(string folderPath)
 		{
 			IUserSettingsService userSettingsService = Ioc.Default.GetRequiredService<IUserSettingsService>();
-			if (userSettingsService.FoldersSettingsService.EnableOverridingFolderPreferences)
+
+			var layout = LayoutPreferences.DefaultLayoutPreferences;
+
+			if (userSettingsService.FoldersSettingsService.EnableOverridingFolderPreferences || userSettingsService.LayoutSettingsService.EnableOverridingSortingPreferences)
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/FolderSettingsViewModel.cs:293]: You're using these two variables two times. One here and one in l.302 and l.323.
You'd be better off saving it inside variables to not access them two times.

Comment (QuaintMako) [inline src/Files.App/ViewModels/FolderSettingsViewModel.cs:293]: You could also reverse the condition of the `if` to avoid a nesting level with an earlier `return layout` after it.

File: src/Files.App/ViewModels/FolderSettingsViewModel.cs [hunk 2]
```diff
@@ -348,6 +360,32 @@ public static void SetLayoutPreferencesForPath(string folderPath, LayoutPreferen
...
+
+			if (!userSettingsService.LayoutSettingsService.EnableOverridingSortingPreferences)
+			{
+				// Do not save OriginalPath as global sort option (only works in recycle bin)
+				if (prefs.DirectorySortOption != SortOption.OriginalFolder &&
+					prefs.DirectorySortOption != SortOption.DateDeleted &&
+					prefs.DirectorySortOption != SortOption.SyncStatus)
+				{
+					userSettingsService.LayoutSettingsService.DefaultDirectorySortOption = prefs.DirectorySortOption;
+				}
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/FolderSettingsViewModel.cs:372]: One liner `if`s should not have curly brackets.

File: src/Files.App/ViewModels/FolderSettingsViewModel.cs [hunk 3]
```diff
@@ -348,6 +360,32 @@ public static void SetLayoutPreferencesForPath(string folderPath, LayoutPreferen
...
+				if (prefs.DirectoryGroupOption != GroupOption.OriginalFolder &&
+					prefs.DirectoryGroupOption != GroupOption.DateDeleted &&
+					prefs.DirectoryGroupOption != GroupOption.FolderPath &&
+					prefs.DirectoryGroupOption != GroupOption.SyncStatus)
+				{
+					userSettingsService.LayoutSettingsService.DefaultDirectoryGroupOption = prefs.DirectoryGroupOption;
+				}
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/FolderSettingsViewModel.cs:379]: One liner `if`s should not have curly brackets.

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs [hunk 4]
```diff
@@ -112,6 +112,9 @@ public PreferencesViewModel()
 
 			AddDateTimeOptions();
 			SelectedDateTimeFormatIndex = (int)Enum.Parse(typeof(DateTimeFormats), DateTimeFormat.ToString());
+			SelectedDefaultSortingIndex = DefaultSortOption == SortOption.FileTag ? 5 : (int)DefaultSortOption;
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs:115]: Where does the `5` comes from? Maybe you could put it inside a well named variable to avoid the magic constant effect?

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs [hunk 5]
```diff
@@ -112,6 +112,9 @@ public PreferencesViewModel()
...
+			SelectedDefaultSortDirectionIndex = (int)DefaultSortDirection;
+			SelectedDefaultGroupingIndex = DefaultGroupOption == GroupOption.FileTag ? 6 : (int)DefaultGroupOption;
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs:117]: Same as the previous point about the `6`.

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs [hunk 6]
```diff
@@ -475,6 +491,87 @@ public bool ListAndSortDirectoriesAlongsideFiles
...
+		private int selectedDefaultSortingIndex;
+		public int SelectedDefaultSortingIndex
+		{
+			get => selectedDefaultSortingIndex;
+			set
+			{
+				if (SetProperty(ref selectedDefaultSortingIndex, value))
+				{
+					OnPropertyChanged(nameof(SelectedDefaultSortingIndex));
+					DefaultSortOption = value < 5 ? (SortOption)value : SortOption.FileTag;
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs:503]: Same as above about the magic constant. If the `5` here has the same use as line 115, putting it inside a `private readonly int` field could solve the issue quite nicely.

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs [hunk 7]
```diff
@@ -475,6 +491,87 @@ public bool ListAndSortDirectoriesAlongsideFiles
...
+		private int selectedDefaultGroupingIndex;
+		public int SelectedDefaultGroupingIndex
+		{
+			get => selectedDefaultGroupingIndex;
+			set
+			{
+				if (SetProperty(ref selectedDefaultGroupingIndex, value))
+				{
+					OnPropertyChanged(nameof(SelectedDefaultGroupingIndex));
+					DefaultGroupOption = value < 6 ? (GroupOption)value : GroupOption.FileTag;
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs:531]: `6` has the name problem and possible solution as `5` on last point.

**Review by QuaintMako — COMMENTED** [review]
LGTM outside of those small propositions to make the code cleaner.

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs (hunk 4 above)
Comment (manfromarce) [reply src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs]: This is a workaround to hide "sync status" and other options which only work in cloud folders or recycle bin. I tried swapping sync status and tag in SortOption and GroupOption, but it should be done in toolbar and context menu too, otherwise the wrong option is selected there.

**Comment by manfromarce** [comment]
The latest commit can be reverted if we prefer to leave the order as is.

File: src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs [hunk 8]
```diff
@@ -207,6 +207,25 @@ public bool CalculateFolderSizes
 			get => Get(false);
 			set => Set(value);
 		}
+
+		public SortDirection DefaultDirectorySortDirection
```
Comment (yaira2) [inline src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:211]: We don't plan on this option for now

File: src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs (hunk 4 above)
Comment (QuaintMako) [reply src/Files.App/ViewModels/SettingsViewModels/PreferencesViewModel.cs]: If it's a work around, I again highly suggest using a named variable to explain its use. Or at least a comment next to it.

File: src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs (hunk 8 above)
Comment (manfromarce) [reply src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:211]: Ok, I will remove it

**Review by QuaintMako — COMMENTED** [review]
Aside of the comments still up, LGTM in general code-wise.

File: src/Files.App/Helpers/ContextFlyoutItemHelper.cs [hunk 9]
```diff
@@ -292,6 +284,14 @@ public static List<ContextMenuFlyoutItemViewModel> GetBaseLayoutMenuItems(Curren
...
 						},
 						new ContextMenuFlyoutItemViewModel()
+						{
+							Text = "SyncStatus".GetLocalizedResource(),
+							IsChecked = itemViewModel.IsSortedBySyncStatus,
+							Command = new RelayCommand(() => itemViewModel.IsSortedBySyncStatus = true),
+							ShowItem = currentInstanceViewModel.IsPageTypeCloudDrive,
+							ItemType = ItemType.Toggle
+						},
+						new ContextMenuFlyoutItemViewModel()
```
Comment (yaira2) [inline src/Files.App/Helpers/ContextFlyoutItemHelper.cs:294]: Why was this changed?

File: src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs (hunk 8 above)
Comment (yaira2) [reply src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:211]: It should be kept in LayoutSettingsService but not here as we don't want to expose the setting.

Comment (yaira2) [inline src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:211]: We shouldn't be exposing this to users right now

File: src/Files.Shared/Enums/SortOption.cs [hunk 10]
```diff
@@ -7,9 +7,9 @@ public enum SortOption : byte
 		DateCreated,
 		Size,
 		FileType,
-		SyncStatus,
 		FileTag,
-		OriginalFolder,
-		DateDeleted
+		SyncStatus, // Cloud drive
```
Comment (yaira2) [inline src/Files.Shared/Enums/SortOption.cs:11]: Why were these reordered?

Comment (manfromarce) [reply src/Files.Shared/Enums/SortOption.cs:11]: This was to separate the options available for all folders from those specific to recycle bin or cloud drives (which need to be hidden from global settings), but I can revert this changes as well as those about default sort direction

File: src/Files.App/Helpers/ContextFlyoutItemHelper.cs (hunk 9 above)
Comment (manfromarce) [reply src/Files.App/Helpers/ContextFlyoutItemHelper.cs:294]: This was necessary after reordering the enum (please see below)

File: src/Files.Shared/Enums/SortOption.cs (hunk 10 above)
Comment (yaira2) [reply src/Files.Shared/Enums/SortOption.cs:11]: That would be preferred 👍

**Comment by manfromarce** [comment]
I should be done with the requested changes.

File: src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs [hunk 11]
```diff
@@ -207,6 +207,19 @@ public bool CalculateFolderSizes
...
 			set => Set(value);
 		}
+
+		public SortOption DefaultDirectorySortOption
+		{
+			get => (SortOption)Get((long)SortOption.Name);
+			set => Set((long)value);
+		}
+
+		public GroupOption DefaultDirectoryGroupOption
```
Comment (yaira2) [inline src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:217]:
```suggestion
		public GroupOption DefaultGroupOption
```

File: src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs [hunk 12]
```diff
@@ -207,6 +207,19 @@ public bool CalculateFolderSizes
 			get => Get(false);
 			set => Set(value);
 		}
+
+		public SortOption DefaultDirectorySortOption
```
Comment (yaira2) [inline src/Files.App/ServicesImplementation/Settings/FoldersSettingsService.cs:211]:
```suggestion
		public SortOption DefaultSortOption
```

File: src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs [hunk 13]
```diff
@@ -239,6 +244,60 @@ public bool CalculateFolderSizes
...
+		private int selectedDefaultGroupingIndex;
+		public int SelectedDefaultGroupingIndex
+		{
+			get => selectedDefaultGroupingIndex;
+			set
+			{
+				if (SetProperty(ref selectedDefaultGroupingIndex, value))
+				{
+					OnPropertyChanged(nameof(SelectedDefaultGroupingIndex));
+					DefaultGroupOption = value == FileTagGroupingIndex ? GroupOption.FileTag : (GroupOption)value;
```
Comment (yaira2) [inline src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:270]:
```suggestion
					FoldersSettingsService.DefaultDirectorySortOption = value == FileTagGroupingIndex ? GroupOption.FileTag : (GroupOption)value;
```
```suggestion
					DefaultGroupOption = value == FileTagGroupingIndex ? GroupOption.FileTag : (GroupOption)value;
```
If we do this, we can remove the `DefaultSortOption` property from line 275.

Comment (manfromarce) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:270]: You probably meant FoldersSettingsService.DefaultGroupOption

Comment (yaira2) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs]: Yeah

Comment (yaira2) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs]: Can you do the same for Sorting?

File: src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs [hunk 14]
```diff
@@ -10,6 +10,9 @@ public class FoldersViewModel : ObservableObject
 	{
 		private IUserSettingsService UserSettingsService { get; } = Ioc.Default.GetRequiredService<IUserSettingsService>();
 
+		private readonly int FileTagSortingIndex = 5;
+		private readonly int FileTagGroupingIndex = 6;
```
Comment (yaira2) [inline src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:14]: What are these values for?

File: src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs (hunk 13 above)
Comment (manfromarce) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs]: Yes

File: src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs (hunk 14 above)
Comment (manfromarce) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:14]: They are used to select the correct option in the comboboxes. Sync status, original path and delete date are hidden because they aren't available for regular folders, so if Tags is set as default the casting would return an invalid index. At first I swapped FileTag and SyncStatus but that enum is referenced from many places.

Comment (yaira2) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:14]: Can you add a comment to the code?

Comment (manfromarce) [reply src/Files.App/ViewModels/SettingsViewModels/FoldersViewModel.cs:14]: Yes, I'll do it

**Review by yaira2 — APPROVED** [review]
LGTM

File: src/Files.App/Strings/en-US/Resources.resw [hunk 15]
```diff
@@ -1,17 +1,17 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!--
-    Microsoft ResX Schema
-
+  <!-- 
+    Microsoft ResX Schema 
+
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:5]:
```suggestion
  <!--
    Microsoft ResX Schema

```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 16]
```diff
@@ -1,17 +1,17 @@
...
+    
+    The primary goals of this format is to allow a simple XML format 
+    that is mostly human readable. The generation and parsing of the 
+    various data types are done through the TypeConverter classes 
     associated with the data types.
-
+    
     Example:
-
+
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:14]:
```suggestion

```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 17]
```diff
@@ -26,36 +26,36 @@
         <value>[base64 mime encoded string representing a byte array form of the .NET Framework object]</value>
         <comment>This is a comment</comment>
     </data>
-
-    There are any number of "resheader" rows that contain simple
+
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:29]:
```suggestion

```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 18]
```diff
@@ -26,36 +26,36 @@
         <value>[base64 mime encoded string representing a byte array form of the .NET Framework object]</value>
         <comment>This is a comment</comment>
     </data>
-
-    There are any number of "resheader" rows that contain simple
+                
+    There are any number of "resheader" rows that contain simple
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:30]:
```suggestion
    There are any number of "resheader" rows that contain simple
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 19]
```diff
@@ -26,36 +26,36 @@
...
-
-    Each data row contains a name, and value. The row also contains a
-    type or mimetype. Type corresponds to a .NET class that support
-    text/value conversion through the TypeConverter architecture.
-    Classes that don't support this are serialized and stored with the
+    
+    Each data row contains a name, and value. The row also contains a 
+    type or mimetype. Type corresponds to a .NET class that support 
+    text/value conversion through the TypeConverter architecture. 
+    Classes that don't support this are serialized and stored with the
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:36]:
```suggestion

    Each data row contains a name, and value. The row also contains a
    type or mimetype. Type corresponds to a .NET class that support
    text/value conversion through the TypeConverter architecture.
    Classes that don't support this are serialized and stored with the
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 20]
```diff
@@ -26,36 +26,36 @@
...
+    type or mimetype. Type corresponds to a .NET class that support 
+    text/value conversion through the TypeConverter architecture. 
+    Classes that don't support this are serialized and stored with the 
     mimetype set.
-
-    The mimetype is used for serialized objects, and tells the
-    ResXResourceReader how to depersist the object. This is currently not
+    
+    The mimetype is used for serialized objects, and tells the 
+    ResXResourceReader how to depersist the object. This is currently not
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:40]:
```suggestion

    The mimetype is used for serialized objects, and tells the
    ResXResourceReader how to depersist the object. This is currently not
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 21]
```diff
@@ -26,36 +26,36 @@
...
     extensible. For a given mimetype the value must be set accordingly:
-
-    Note - application/x-microsoft.net.object.binary.base64 is the format
-    that the ResXResourceWriter will generate, however the reader can
+    
+    Note - application/x-microsoft.net.object.binary.base64 is the format 
+    that the ResXResourceWriter will generate, however the reader can 
     read any of the formats listed below.
-
+
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:46]:
```suggestion

```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 22]
```diff
@@ -26,36 +26,36 @@
...
     read any of the formats listed below.
-
+    
     mimetype: application/x-microsoft.net.object.binary.base64
-    value   : The object must be serialized with
+    value   : The object must be serialized with 
             : System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
             : and then encoded with base64 encoding.
-
+
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:51]:
```suggestion

```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 23]
```diff
@@ -26,36 +26,36 @@
...
     mimetype: application/x-microsoft.net.object.binary.base64
-    value   : The object must be serialized with
+    value   : The object must be serialized with 
             : System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
             : and then encoded with base64 encoding.
-
+    
     mimetype: application/x-microsoft.net.object.soap.base64
-    value   : The object must be serialized with
+    value   : The object must be serialized with
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:53]:
```suggestion
    value   : The object must be serialized with
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 24]
```diff
@@ -26,36 +26,36 @@
...
+    
     mimetype: application/x-microsoft.net.object.soap.base64
-    value   : The object must be serialized with
+    value   : The object must be serialized with 
             : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
             : and then encoded with base64 encoding.
 
     mimetype: application/x-microsoft.net.object.bytearray.base64
-    value   : The object must be serialized into a byte array
+    value   : The object must be serialized into a byte array
```
Comment (QuaintMako) [inline src/Files.App/Strings/en-US/Resources.resw:58]:
```suggestion
    value   : The object must be serialized into a byte array
```

**Review by QuaintMako — APPROVED** [review]
LGTM, aside of the small formatting I suggested above to remove unneeded spaces.

**Review by QuaintMako — APPROVED** [review]
